### ISSUE TITLE
pin-depends: is not transitive during opam install

### ISSUE THREAD

**Issue by Lupus** [ocaml/opam#4003]
@avsm [pointed out](https://github.com/ocaml/dune/issues/2708#issuecomment-539095701) that pin-depends should be resolved transitively and looks like it indeed works despite official documentation claiming this is not supported [1].

I've tried pinning a package, that has some `pin-depends:` in its opam file, and I was suggested to accept additional pinnings required by that package.

When I added the package with `pin-depends:` to my top level package as `pin-depends:`, and ran `opam install . --deps-only --working-dir`, transitive pins did not work, I was not suggested that additional pinings are required.

Is this also expected to work? We use opam install command to fetch package dependencies into local switch both during development and CI runs.

```
# opam-version      2.0.5 
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=debian os-version=10
# solver            builtin-mccs+glpk
```

[1] https://opam.ocaml.org/doc/Manual.html#opamfield-pin-depends

**Comment by Lupus** [comment]
Some packages that we pin are forks of packages, available in public opam repo, how can we ensure that opam picks those forked versions from our internal or github repositories?

**Comment by avsm** [comment]
I thought that they were transitive due to the (somewhat cryptic :-) 2.0.2 fix for @hannesm:
>   * upgrade pin depends on pinning

How many things are you forking, @Lupus?  Some of the tooling we are building in https://github.com/avsm/duniverse may work for this usecase.

**Comment by Lupus** [comment]
We try to avoid that, or upstream changes, but sometimes that does not work. Duniverse is just a vendoring tool for dune, right? I've looked at it already, is it capable at solving transitivity problem, when external dependency has some other external dependency, so that both are accounted?

**Comment by AltGr** [comment]
Repositories are [always ordered](https://opam.ocaml.org/doc/Manual.html#Repositories) (higher priority first), which means that if you define the repos as `custom,default` in your switch¹, any version of a package that exists in both will use the definition from your switch.

A new version of a package appearing in the default repository will be seen if it isn't yet overriden in your overlay. If this is a problem, you should use your own fork or branch of the default repository, so that you can control when it gets updates. Keeping it unchanged and having a separate repo with only your specific package definitions is easier for maintainance.

The `opam admin` command has some tools for managing package repositories.

Basically, the remaining constraint is that you need the definitions of all your pins/overlays/packages centralised somewhere, since opam doesn't do dependency discovery. 

Hope this helps!

> NB: just noticed the instructions on `opam repo add` in the manual are outdated, fixing right away.

¹ either with `opam switch create --repos=...` or `opam repo set-repos`

**Comment by Lupus** [comment]
Oh, that is nice! So if we have fork of "foo" in our internal repo at version 1.2.3, and have repos properly configured for a switch, we should never see "foo" version 1.2.5 installed from official repo?

**Comment by Lupus** [comment]
Okay, I managed to hack opam-publish so that it works in automatic publishing mode from Gitlab CI pipelines, with CI token in .netrc and deploy SSH key for local opam-repository push it works fine even with auth-protected repositories. This will even scale to mirroring Github repos to internal Gitlab, and running CI pipelines on them to publish fresh versions by tags faster than they normally appear in official repo! :)

@AltGr, can you please confirm that `opam repo set-repos our-repo default` overrides package definitions so that whatever is found in our-repo takes precedence over what's in default repo (even if default has higher version available)? Thanks!

**Comment by AltGr** [comment]
(sorry for the lag)
No, I am afraid I wasn't clear: if you use `opam repo set-repos our-repo default`, you will "see" any new version appearing in `default` even if another version was overriden in `our-repo`.
To work around that, you would have to use a mirror of default, e.g.:
```
opam repo set-url default git+https://my-clone-of-the-repo
```
so that you can control when updates are applied to the clone. Alternatively, you can just "pin" a commit/branch with:
```
opam repo set-url default git+https://github.com/ocaml/opam-repository#COMMITHASH
```

**Comment by Lupus** [comment]
So what exactly is defined with opam repo order then?

вт, 12 нояб. 2019 г., 11:56 Louis Gesbert <notifications@github.com>:

**Comment by samoht** [comment]
Can you clarify the current state of recursive pinning?

I thought to know that recursive pins (ie, if you depend on a package that has a `pin-depends` dependency, it'll install it too) and the few tests I've done seemed to suggest it indeed works (at least for depth=2 --> you `opam pin` a package that has a `pin-depends` fields --> in that case opam will pin the depencies too).

But after reading that issue, I'm pretty confused as what is the expected behavior of opam now.

**Comment by kit-ty-kate** [comment]
> I thought to know that recursive pins (ie, if you depend on a package that has a `pin-depends` dependency, it'll install it too)

That definition doesn't match the current spec: https://opam.ocaml.org/doc/Manual.html#opamfield-pin-depends

Basically, if you're explicitly pinning a package (be it with `opam pin` or `opam install`), that package's `pin-depends` will be taken into account. Otherwise it is ignored.

> and the few tests I've done seemed to suggest it indeed works (at least for depth=2 --> you `opam pin` a package that has a `pin-depends` fields --> in that case opam will pin the depencies too).

I'm not exactly sure what you mean by "in that case opam will pin the depencies too" but i can't reproduce the depth=2 (if i understand what you mean by "depth").

For example with the following local opam file:
```
opam-version: "2.0"
name: "test"
depends: "bob"
pin-depends: ["bob.dev" "git+https://github.com/dinosaure/bob.git#acc99beaaab497a69c55b296ac7f850d3cb0033c"]
```
Calling `opam pin add .` will pin `bob` but **not** `bob`'s pin-depends (`spoke`)

**Comment by samoht** [comment]
> Calling opam pin add . will pin bob but not bob's pin-depends (spoke)

Ok I'm very confused then. I always assumed opam was doing this. Why are we doing this? ie. doing recursive pinning with depth<=1 ?

**Comment by kit-ty-kate** [comment]
The reason is described just above by @AltGr in https://github.com/ocaml/opam/issues/4003#issuecomment-539528783 (the feature was originally added in https://github.com/ocaml/opam/pull/3059)
This can of course be reevaluated (that's why this ticket hasn't been closed) but it seems to be a rather tricky design problem on first glance
